### ISSUE TITLE
Multipart uploads with S3 pre-signed URLs

### ISSUE THREAD

**Issue by Andy6Loda** [aws/aws-sdk-cpp#1548]
Is your feature request related to a problem? Please describe.
1.Segment upload requires S3 multipart upload, for that i need to generate presigned URLs for each S3 multipart operation. But i can not find it on asw-sdk-cpp just like https://www.altostra.com/blog/multipart-uploads-with-s3-presigned-url

Describe the solution you'd like
Could you tell me how i can create pre-signed URLs for each part?

**Comment by KaibaLopez** [comment]
Hi @Andy6Loda ,

So, not really a full sample, but the code to generate a presigned url (for any request on s3) would look similar to something like: 
```cpp
Aws::Client::ClientConfiguration clientConfig("default");
auto s3Client = Aws::MakeShared<Aws::S3::S3Client>("s3Client", clientConfig);
UploadPartRequest partRequest;
partRequest.WithBucket("").WithKey("");
...
std::cout<< "Presigned for part: " << s3Client->GeneratePresignedUrl(partRequest.GetBucket(),partRequest.GetKey(),Aws::Http::HttpMethod::HTTP_GET) << std::endl;
```
Hope this helps.

**Comment by Andy6Loda** [comment]
Hi @KaibaLopez ,
I think maybe i do not describe it clear：

Business scenarios:
     Server：one admin, well create pre-signed URLs for each part
     Client :  multiple user, have file's content, file total size large than 1TB， segment-upload in parallel
so, Client-user need every part  pre-signed URLs

After read you words, i think the way you said is:
_step-1:multipartUpload file_
     a."CreateMultipartUploadRequest" get the upload_id;
     b."UploadPartRequest" get the Etag with part num;
     c."CompleteMultipartUploadRequest" make sure the file trans over.
_step-2:GeneratePresignedUrl_
     a."GeneratePresignedUrl"  the whole file

But, i expect the final process looks like this:
    a.Admin : Initiate a multipart upload
    b.Admin : Create pre-signed URLs for each part
    c.User : Upload the object’s parts  _//the user donot care where is storage and authentication_
    d.Admin : Complete multipart upload

Thanks for the explanation,  I still not found way solve it ！

**Comment by KaibaLopez** [comment]
Hi @Andy6Loda ,
Not really, I guess I was just too specific in my response, building the code is really up to you, we won't support that out of the box.
What I posted was a simple call to generate a presigned url for a single upload part, which I assumed was all you needed to build the code.
If you need a more complete sample, best I can do is point you towards the transfer manager's [multipart upload](https://github.com/aws/aws-sdk-cpp/blob/75b4ece672634b81879baf75d9103cd106a972ab/aws-cpp-sdk-transfer/source/transfer/TransferManager.cpp#L273), it won't have the specific code you need but it's a good base on how to create a multi part upload and handling each part.

**Comment by Andy6Loda** [comment]
Thank you very much for your reply @KaibaLopez 

I solved my needs in the following ways：

```
Aws::Client::ClientConfiguration clientConfig("default");
auto s3Client = Aws::MakeShared<Aws::S3::S3Client>("s3Client", clientConfig);

_//step 1   :Initiate a multipart upload_
Aws::S3::Model::CreateMultipartUploadRequest request;
request.WithBucket(bucket.data()).WithKey(key.data());
m_client->CreateMultipartUpload(request);  //you can get the upload_id

_//step 2-1: set the part of target;_
Aws::S3::Model::UploadPartRequest partRequest;
partRequest.SetBucket("...");
partRequest.SetKey("...");
partRequest.SetUploadId(upload_id);
partRequest.SetPartNumber(xxx);//This is a positive integer between 1 and 10,000
_//step 2-2: Create pre-signed URLs for each part_
 m_client->**AWSClient::**GeneratePresignedUrl(partRequest, url, Aws::Http::HttpMethod::HTTP_PUT,
                                                             Aws::Http::QueryStringParameterCollection(),
                                                             expirationInSeconds);
    //Attention：the expirationInSeconds can not use default value ， otherwise， it will be alarm expire when you upload
   //                    the url format :  http://ip:port/bucket/keys   used by  MinIO
  //                     this return string  is  pre-signed URLs  for one part

_//step 3   :  upload content by http-client(libevent) or postman-test or your  fancy_
。。。。  // here you will get the HTTP respones  header 'ETag'

_//step 4  :   Complete multipart upload_
Aws::S3::Model::CompleteMultipartUploadRequest request;
  request.WithBucket(bucket.data()).WithKey(key.data()).WithUploadId("4f3d7f64-8eb4-4336-80da-36416281f04a");
  Aws::S3::Model::CompletedMultipartUpload completedMultipartUpload;
  Aws::S3::Model::CompletedPart completedPart;
//Pay attention to the matching relationship
  completedPart.WithPartNumber(1).WithETag("1891b2fea417c0a4d81c8735961aa5e0");
completedMultipartUpload.AddParts(completedPart);
  request.SetMultipartUpload(completedMultipartUpload);
m_client->CompleteMultipartUpload(request);
```

Unfortunately, I don't think it's friendly enough here ： m_client->AWSClient::GeneratePresignedUrl

Why do not open this func ?

**Comment by KaibaLopez** [comment]
I don't understand the question, what do you mean open the func? 
Also, the AWSClient namespace is not necessary for that call, `m_client->GeneratePresignedUrl(` should be ok.
Anyways, glad you manage to figure out the solution for your problem.

**Comment by Andy6Loda** [comment]
![image](https://user-images.githubusercontent.com/71309695/103397703-7075f400-4b74-11eb-9239-b9a1fa80de8a.png)
If the function of the derived class and the function of the base class have the same name, but the parameters are different, then the function of the base class will be hidden whether there is virtual or not.

**Comment by KaibaLopez** [comment]
yea ok I see where I was wrong there, but still, I don't think this would be something we'd change considering it is not the "expected" usage... And to be clear here, I'm not saying it's something we wouldn't consider, just that it would be very low on the priority list, we'd still take a PR though.
For now, and for the sake of keeping clarity on the issue, I will be closing this as it seems to me like no guidance is required, feel free to open a new issue as a feature request.
